streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +39,682 @@
+
+        state = ChangelogReaderState.STANDBY_UPDATING;
+    }
+
+    /**
+     * Since it is shared for multiple tasks and hence multiple state managers, the registration would take its
+     * corresponding state manager as well for restoring.
+     */
+    @Override
+    public void register(final TopicPartition partition, final ProcessorStateManager stateManager) {
+        final ChangelogMetadata changelogMetadata = new ChangelogMetadata(partition, stateManager);
+
+        // initializing limit offset to 0L for standby changelog to effectively disable any restoration until it is updated
+        if (state == ChangelogReaderState.STANDBY_UPDATING && stateManager.changelogAsSource(partition)) {
+            changelogMetadata.restoreLimitOffset = 0L;
         }
 
-        try {
-            final ConsumerRecords<byte[], byte[]> records = restoreConsumer.poll(pollTime);
-
-            for (final TopicPartition partition : needsRestoring) {
-                final StateRestorer restorer = stateRestorers.get(partition);
-                final long pos = processNext(records.records(partition), restorer, restoreToOffsets.get(partition));
-                restorer.setRestoredOffset(pos);
-                if (restorer.hasCompleted(pos, restoreToOffsets.get(partition))) {
-                    restorer.restoreDone();
-                    restoreToOffsets.remove(partition);
-                    completedRestorers.add(partition);
-                }
-            }
-        } catch (final InvalidOffsetException recoverableException) {
-            log.warn("Restoring StreamTasks failed. Deleting StreamTasks stores to recreate from scratch.", recoverableException);
-            final Set<TopicPartition> partitions = recoverableException.partitions();
-            for (final TopicPartition partition : partitions) {
-                final StreamTask task = active.restoringTaskFor(partition);
-                log.info("Reinitializing StreamTask {} for changelog {}", task, partition);
-
-                needsInitializing.remove(partition);
-                needsRestoring.remove(partition);
-
-                final StateRestorer restorer = stateRestorers.get(partition);
-                restorer.setCheckpointOffset(StateRestorer.NO_CHECKPOINT);
-                task.reinitializeStateStoresForPartitions(recoverableException.partitions());
-            }
-            restoreConsumer.seekToBeginning(partitions);
+        if (changelogs.putIfAbsent(partition, changelogMetadata) != null) {
+            throw new IllegalStateException("There is already a changelog registered for " + partition +
+                ", this should not happen: " + changelogs);
         }
+    }
 
-        needsRestoring.removeAll(completedRestorers);
+    private ChangelogMetadata restoringChangelogByPartition(final TopicPartition partition) {
+        final ChangelogMetadata changelogMetadata = changelogs.get(partition);
```

Comment:
nit: I'm in favor of doing the null check in first time register

Reply:
I think it is a bit orthogonal: here we poll a bunch of records organized by partitions, and we check that each partition is indeed in the registered changelogs, if the code is bug-free it should always be true.